brave/src/main/java/brave/internal/baggage/ExtraFactory.java

```diff
@@ -0,0 +1,188 @@
+import brave.propagation.TraceContextOrSamplingFlags;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * This manages an mutable element of {@link TraceContext#extra()} of type ({@link Extra}). The most
+ * common usage is assigning {@linkplain BaggageField baggage} parsed from the request to a span.
+ * However, this supports other use cases such as extended trace contexts and secondary sampling.
+ *
+ * <h3>Method overview</h3>
+ * {@link #decorate(TraceContext)} ensures that the resulting {@link TraceContext} has an instance
+ * of "extra" ({@link #<E>}), that it is associated only with the context's span ID, and that fields
+ * inside "extra" considers upstream values, such as from a parent span.
+ *
+ * <p>{@link #create()} is used to add an instance of {@link #<E>} to an {@link
+ * TraceContextOrSamplingFlags extraction result}. Concretely, this allows additional request data,
+ * such as a request ID, to be inherited from a request, and possibly changed, before a trace
+ * context is {@linkplain #decorate(TraceContext) decorated}.
+ *
+ * <h3>Integration</h3>
+ * The only valid way to integrate this is via a custom {@link Propagation.Factory}. Its {@link
+ * Propagation.Factory#decorate(TraceContext)} method MUST dispatch to {@link
+ * #decorate(TraceContext)}.
+ *
+ * <p>If request state is needed, {@link Propagation#extractor(Propagation.Getter)}
+ * MUST call {@link #create()} only once per request. The result should be added to
+ * TraceContextOrSamplingFlags.Builder#addExtra(Object)}.
+ *
+ * <h3>Notes</h3>
+ * <p>The reason {@link #create()} is decoupled from {@link #decorate(TraceContext)} is because
+ * extra data can exist in a request even if a trace context does not. Even when a trace context
+ * exists in headers, it is only used directly when {@linkplain Tracing.Builder#supportsJoin(boolean)
+ * join is supported}. In other words, {@link #<E>} carries any data from the request until it can
+ * be associated with a {@link TraceContext} (via {@link #decorate(TraceContext)}.
+ *
+ * <p>It may not be intuitive why implementations should always add
+ *
+ * <p>If your data is not mutable, and is constant through the trace, do not use this. Instead, add
+ * your own type via {@link TraceContextOrSamplingFlags.Builder#addExtra(Object)} as the tracer will
+ * copy it down to children by default.
+ *
+ * <p>Even though {@link #<E>} is mutable, it must be copy-on-write internally. This type manages
+ * state forking that ensures that updates to child spans are invisible to their parents or
+ * siblings.
+ *
+ * @param <E> They type of {@link Extra} managed by this factory. Must be a final class. The type is
+ * typically package private to avoid accidental interference.
+ * @param <F> An instance of this factory. {@link #<E>} should be associated with only one factory.
+ */
+public abstract class ExtraFactory<E extends Extra<E, F>, F extends ExtraFactory<E, F>> {
```

Comment:
because I added so much javadoc, this became a new file. The logic hasn't changed except coersing bugs to the input context